File: webgl-ann-experiment/src/com/sw_engineering_candies/example/shared/Neuron.java
/**
 * Copyright (C) 2013, Markus Sprunck
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or
 * without modification, are permitted provided that the following
 * conditions are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above
 *   copyright notice, this list of conditions and the following
 *   disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * - The name of its contributor may be used to endorse or promote
 *   products derived from this software without specific prior
 *   written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.sw_engineering_candies.example.shared;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.google.common.base.Preconditions;

public class Neuron {

	public enum Type {
		INPUT, OUTPUT, INNER
	}

	private final List<Link> links = new ArrayList<Link>();

	private double output = 0.0;

	private double outputExpected = 0.0;

	private double outputError = 0.0;

	private double input = 0.0;

	private double outputDerived = 1.0;

	private final Type type;

	public Neuron() {
		this(Type.INNER);
	}

	public Neuron(final Type type) {
		this.type = type;
	}

	public void createLink(final Neuron source, final double weight) {
		Link link = null;

		Preconditions.checkNotNull(source);
		Preconditions.checkArgument(!source.equals(this));

		link = new Link();
		link.setSource(source);
		link.setWeight(weight);
		links.add(link);
	}

	private double functionFermi(final double x) {
		if (x > 15) {
			return 1.0;
		} else if (x < -15) {
			return 0.0;
		} else {
			final double y = 1.0 / (1.0 + Math.exp(-x));
			return y;
		}
	}

	private double functionFermiDerive(final double x) {
		final double z = functionFermi(x);
		final double y = z * (1 - z);
		return y;
	}

	public double getInput() {
		return input;
	}

	public List<Link> getLinks() {
		return Collections.unmodifiableList(links);
	}

	public double getOutput() {
		return output;
	}

	public double getOutputDerived() {
		return outputDerived;
	}

	public double getOutputExpected() {
		return outputExpected;
	}

	public boolean isInputNeuron() {
		return Type.INPUT.equals(type);
	}

	public boolean isInnerNeuron() {
		return Type.INNER.equals(type);
	}

	public boolean isOutputNeuron() {
		return Type.OUTPUT.equals(type);
	}

	public void setInput(final double input) {
		Preconditions.checkState(Type.INPUT == type);
		this.input = input;
	}

	public void recall() {

		if (Type.INPUT.equals(type)) {
			output = input;
			outputDerived = 1.0;
		} else {
			double sum = 0.0;
			for (final Link link : links) {
				sum += link.getSource().getOutput() * link.getWeight();
			}
			output = functionFermi(sum);
			outputDerived = functionFermiDerive(sum);
		}
	}

	public void calculateEvaluateOutputError() {
		Preconditions.checkArgument(Type.OUTPUT == type);
		outputError = output - outputExpected;
	};

	public void calculateEvaluateOutputErrorHiddenNeurons(
			final double m_FlatSpot) {
		Preconditions.checkArgument(Type.INPUT != type);

		for (final Link link : links) {
			final double derivation = outputDerived + m_FlatSpot;
			final double oldError = link.getSource().getOutputError();
			link.getSource().setOutputError(
					derivation * outputError * link.getWeight() + oldError);
		}
	};

	public void setOutputError(final double outputError) {
		this.outputError = outputError;
	}

	public double getOutputError() {
		return outputError;
	}

	public void setOutputExpected(final double outputExpected) {
		this.outputExpected = outputExpected;
	}

	@Override
	public String toString() {
		return "{\"y\":" + output + ",\"y_ex\":" + outputExpected + "}";
	}

}

File: typescript/src/Neuron.ts
import {Link} from './Link';

export enum NeuronType { INPUT = 'INPUT', OUTPUT = 'OUTPUT', INNER = 'INNER' }

export class Neuron {
    links: Link[] = [];
    output = 0.0;
    outputExpected = 0.0;
    outputError = 0.0;
    input = 0.0;
    outputDerived = 1.0;
    type: NeuronType;

    constructor(type: NeuronType = NeuronType.INNER) {
        this.type = type;
    }

    createLink(source: Neuron, weight: number) {
        if (!source) throw new Error('source is null');
        if (source === this) throw new Error('source equals this');
        const link = new Link();
        link.source = source;
        link.weight = weight;
        this.links.push(link);
    }

    getInput() {
        return this.input;
    }

    getLinks(): Link[] {
        return this.links;
    }

    getOutput() {
        return this.output;
    }

    getOutputDerived() {
        return this.outputDerived;
    }

    getOutputExpected() {
        return this.outputExpected;
    }

    isInputNeuron() {
        return this.type === NeuronType.INPUT;
    }

    isInnerNeuron() {
        return this.type === NeuronType.INNER;
    }

    isOutputNeuron() {
        return this.type === NeuronType.OUTPUT;
    }

    setInput(v: number) {
        if (this.type !== NeuronType.INPUT) throw new Error('IllegalState: not input neuron');
        this.input = v;
    }

    recall() {
        if (this.type === NeuronType.INPUT) {
            this.output = this.input;
            this.outputDerived = 1.0;
        } else {
            let sum = 0.0;
            for (const link of this.links) {
                // use getter to avoid direct property access on linked neuron
                sum += (link.source!.getOutput()) * link.weight;
            }
            this.output = this.functionFermi(sum);
            this.outputDerived = this.functionFermiDerive(sum);
        }
    }

    calculateEvaluateOutputError() {
        if (this.type !== NeuronType.OUTPUT) throw new Error('IllegalArgument: not output');
        this.outputError = this.output - this.outputExpected;
    }

    calculateEvaluateOutputErrorHiddenNeurons(m_FlatSpot: number) {
        if (this.type === NeuronType.INPUT) throw new Error('IllegalArgument: is input');
        for (const link of this.links) {
            const derivation = this.outputDerived + m_FlatSpot;
            const oldError = link.source!.getOutputError();
            link.source!.setOutputError(derivation * this.outputError * link.weight + oldError);
        }
    }

    setOutputError(v: number) {
        this.outputError = v;
    }

    getOutputError() {
        return this.outputError;
    }

    setOutputExpected(v: number) {
        this.outputExpected = v;
    }

    toString() {
        return `{"y":${this.output},"y_ex":${this.outputExpected}}`;
    }

    private functionFermi(x: number) {
        if (x > 15) return 1.0;
        if (x < -15) return 0.0;
        const y = 1.0 / (1.0 + Math.exp(-x));
        return y;
    }

    private functionFermiDerive(x: number) {
        const z = this.functionFermi(x);
        return z * (1 - z);
    }
}
